### PR TITLE
Add support for performance_schema

### PR DESCRIPTION
**Pull request by SuperQ**
Add support for getting table metrics from MySQL
`performance_schema.table_io_waits_summary_by_table`

### CODE REVIEW

**Comment by beorn7** [comment]
What's your suggested course of action now?

- Get this submitted ASAP and refactor it to a saner state later?
- Go through a detailed review now to improve things?
- Let me write additional commits to straighten thing (also the legacy ones) up with this PR?

**Comment by SuperQ** [comment]
So I think some refactoring needs to happen in-place with regards to the labels.  I tested the code, it does not work as-is.

**Comment by beorn7** [comment]
I'll have a closer look ASAP. It sounds we should first clean up the code because that will make debugging easier.

**Comment by beorn7** [comment]
@SuperQ The latest commit brings things in line with sanity. However, I haven't even tried it out, and obviously, we should have a test reading from some example input. Unfortunately, I'm out of time. Perhaps I'll find some time during my vacation, but unlikely... You might want to try it out in the meantime...

File: mysqld_exporter.go [hunk 1]
```diff
@@ -12,297 +12,326 @@ import (
...
-	}
-
-	e.commands.Describe(ch)
-	e.connectionErrors.Describe(ch)
-	e.innodbRows.Describe(ch)
-	e.performanceSchema.Describe(ch)
+	// We cannot know in advance what metrics the exporter will generate
+	// from MySQL. So we use the poor man's describe method: Run a collect
+	// and send the descriptors of all the collected metrics. The problem
+	// here is that we need to connecte to the MySQL DB. If it is currently
```
Comment (juliusv) [inline mysqld_exporter.go]: s/connecte/connect/

File: mysqld_exporter.go [hunk 2]
```diff
@@ -12,297 +12,326 @@ import (
...
-	dsn               string
-	mutex             sync.RWMutex
-	duration, error   prometheus.Gauge
-	totalScrapes      prometheus.Counter
-	metrics           map[string]prometheus.Gauge
-	commands          *prometheus.CounterVec
-	connectionErrors  *prometheus.CounterVec
-	innodbRows        *prometheus.CounterVec
-	performanceSchema *prometheus.CounterVec
+	mutex           sync.Mutex
```
Comment (grobie) [inline mysqld_exporter.go]: Unused, can go away.

**Comment by SuperQ** [comment]
@beorn7 I fixed a race condition with the help of @juliusv.  I tested it against a real mysqld server and got good results.

**Comment by SuperQ** [comment]
Added a bool flag to make this optional

**Comment by SuperQ** [comment]
Removed [WIP], this is ready for final review. :) :) :)

File: mysqld_exporter.go [hunk 3]
```diff
@@ -1,311 +1,347 @@
...
+	// exported metrics during the runtime of the exporter.
+
+	metricCh := make(chan prometheus.Metric)
+	doneCh := make(chan struct{})
+
+	go func() {
+		for m := range metricCh {
+			ch <- m.Desc()
+		}
+		doneCh <- struct{}{}
```
Comment (fabxc) [inline mysqld_exporter.go]: This works just fine but `close(doneCh)` is the more idiomatic solution.

File: mysqld_exporter.go [hunk 4]
```diff
@@ -1,311 +1,347 @@
...
+var (
+	listenAddress = flag.String(
+		"web.listen-address", ":9104",
+		"Address to listen on for web interface and telemetry.",
+	)
+	metricPath = flag.String(
+		"web.telemetry-path", "/metrics",
+		"Path under which to expose metrics.",
+	)
+	perfTableIoWaits = flag.Bool(
```
Comment (fabxc) [inline mysqld_exporter.go]: `perfTableIOWaits`

**Comment by fabxc** [comment]
Looks good aside from nit comments :+1:

**Comment by SuperQ** [comment]
@juliusv Any additional issues?

File: mysqld_exporter.go [hunk 5]
```diff
@@ -1,311 +1,347 @@
...
+		}
+		doneCh <- struct{}{}
+	}()
 
-	ch <- e.duration.Desc()
-	ch <- e.totalScrapes.Desc()
-	ch <- e.error.Desc()
+	e.Collect(metricCh)
+	close(metricCh)
+	close(doneCh)
```
Comment (juliusv) [inline mysqld_exporter.go]: I think you've misunerstood Fabian here. Line 143 should be `close(doneCh)`, while this line here should still be `<-doneCh`.

**Comment by juliusv** [comment]
The rest looks roughly ok from a brief scan.

**Comment by juliusv** [comment]
:+1: on the channel fix

**Comment by fabxc** [comment]
:shipit: :ship:

**Comment by beorn7** [comment]
I had a few things written for this while on vacation (especially for testability) but never pushed it because of my crappy connectivity.

I'll create a separate PR if it's worth it...

**Pull request merged**